server/src/main/java/org/elasticsearch/cluster/features/NodeFeaturesFixupListener.java

```diff
@@ -0,0 +1,217 @@
+                        builder.putNodeFeatures(e.getKey(), e.getValue());
+                        modified = true;
+                    }
+                }
+                c.success(() -> {});
+            }
+            return modified ? builder.build() : context.initialState();
+        }
+    }
+
+    @Override
+    public void clusterChanged(ClusterChangedEvent event) {
+        if (event.nodesDelta().masterNodeChanged() && event.localNodeMaster()) {
+            /*
+             * Execute this if we have just become master.
+             * Check if there are any nodes that should have features in cluster state, but don't.
+             * This can happen if the master was upgraded from before 8.13, and one or more non-master nodes
+             * were already upgraded. They don't re-join the cluster with the new master, so never get their features
+             * (which the master now understands) added to cluster state.
+             * So we need to do a separate transport call to get the node features and add them to cluster state.
+             * We can't use features to determine when this should happen, as the features are incorrect.
+             * We also can't use transport version, as that is unreliable for upgrades
+             * from versions before 8.8 (see TransportVersionFixupListener).
+             * So the only thing we can use is release version.
+             * This is ok here, as Serverless will never hit this case, so the node feature fetch action will never be called on Serverless.
+             * This whole class will be removed in ES v9.
+             */
+            ClusterFeatures nodeFeatures = event.state().clusterFeatures();
+            Set<String> queryNodes = event.state()
+                .nodes()
+                .stream()
+                .filter(n -> n.getVersion().onOrAfter(Version.V_8_15_0))
+                .map(DiscoveryNode::getId)
+                .filter(n -> getNodeFeatures(nodeFeatures, n).isEmpty())
+                .collect(Collectors.toSet());
+
+            if (queryNodes.isEmpty() == false) {
+                logger.debug("Fetching actual node features for nodes {}", queryNodes);
+                queryNodesFeatures(queryNodes, 0);
+            }
+        }
+    }
+
+    @SuppressForbidden(reason = "Need to access a specific node's features")
+    private static Set<String> getNodeFeatures(ClusterFeatures features, String nodeId) {
+        return features.nodeFeatures().getOrDefault(nodeId, Set.of());
+    }
+
+    private void scheduleRetry(Set<String> nodes, int thisRetryNum) {
+        // just keep retrying until this succeeds
+        logger.debug("Scheduling retry {} for nodes {}", thisRetryNum + 1, nodes);
+        scheduler.schedule(() -> queryNodesFeatures(nodes, thisRetryNum + 1), RETRY_TIME, executor);
+    }
+
+    private void queryNodesFeatures(Set<String> nodes, int retryNum) {
+        // some might already be in-progress
```

Comment:
This is for the case this runs more times, right? What are the cases? 
Retry for failure, but also more nodes joined?
Would it be possible to have a test for this?

Reply:
I've added a test for overlapping operations

Reply:
Thanks!